Expansion of tomato integration

spiritualtheory opened the pull request:
Adds ability to use tomato for helpline and meeting search without breaking admin or local YAP routing.

New config settings :
tomato_helpline_fallback
tomato_meeting_search

Both false by default

pjaudiomv commented:
Hey thanks for the PR, you can actually already do this by adding a query string to your incoming webhook for ex `https://mydomain.com/yap/index.php?override_tomato_helpline_routing=true`

https://github.com/bmlt-enabled/yap/wiki/Tomato-Helpline-Routing

then you would set your root server in your config to `static $bmlt_root_server = "https://tomato.bmltenabled/main_server";`

pjaudiomv commented:
so for local helpline routing to work there is a new feature coming out in next release its in  the latest beta now.

pjaudiomv commented:
you would provision another number and set the incoming webhook to `?override_extension_dial=true`

then set the helpline fields in your root server to that number followed by the service body id `508-555-1212|wwww2#` would automatically forward the call to service body 2

pjaudiomv commented:
I like this solution though, have you tested a bunch of different settings with it.

spiritualtheory commented:
Initial issue/inconvenience : 
Enabling tomato_helpline_routing in config breaks admin panel

Using the override in webhook breaks the local routing - causes an infinite loop of just forwarding the call to the number called since it shows up in tomato as the helpline number

The change here is you can still use admin, and have working YAP volunteer routing AND fallback to tomato routing if the location is somewhere outside of the local config root server.

No need to provision an additional numbe, change webhooks, or put the tomato server as your bmlt root server

Nothing complicated. Just add settings to config of 
tomato_helpline_fallback = true 
and/or
tomato_meeting_search = true 
if you want tomato as the backup to routing calls outside the local config root server.

spiritualtheory commented:
So as a test case in this - 

In using the webhook or config override of tomato_helpline_routing : 
NNY's YAP (315) 847-3842 has 3 areas on YAP volunteer routing that all show that number as their helpline number. Using tomato as the static or override server causes it to keep forwarding to that base number - so it just forwards to itself.

In using the slightly modified function.php that just adds in a function to check for tomato fallback and adjusts the override accordingly - I just set true or false for 2 settings to turn on and off tomato as a fallback if the caller is not being directed to a local service body.

Meeting search is an easy one - there's no reason to have to test for other conditions - if the tomato meeting search is true, it just returns the tomato_url setting

With the phone service, when the phone volunteer option is hit, it hits tomato with a search for the bmlt root server that would be handling the call routing - if that bmlt root server is the same as the one in the config file, then it uses the local config bmlt server and not tomato - so it doesn't interfere with YAP volunteer call routing or the admin panel. If the bmlt root server handling the call routing isn't the same, then it just sets tomato_helpline_routing to true so that all the other functions that need that reference get it. Admin panel doesn't route through this.

Tested with all 4 conditions
meeting search off / helpling fallback off
meeting search off / helpline fallback on
meeting search on / helpline fallback off
meeting search on / helpling fallback on

Tested by text and call. When enabled, was able to successfully route to Lexington KY as well as the local 3 areas of YAP volunteer routing without issue. Also when enabled, was able to find meetings anywhere. When disabled, behavior was the same as just not having any tomato access - could only see and be routed to what was visible on the local root server.

pjaudiomv commented:
Awesome this seems like a nice easier solution, Danny is the lead maintainer so would like to hear him weigh in before approving.

radius314 commented on endpoints/_includes/functions.php:
```diff
@@ -746,6 +748,30 @@ function helplineSearch($latitude, $longitude)
+	$helpline_search_radius=setting('helpline_search_radius');
+	if ($search_state=="1") {
+		if (setting('tomato_helpline_fallback')) {
+			$bmlt_search_endpoint=setting('tomato_url')."/client_interface/json/?switcher=GetSearchResults&data_field_key=longitude,latitude,root_server_uri&sort_results_by_distance=1&long_val={LONGITUDE}&lat_val={LATITUDE}&geo_width=" . $helpline_search_radius;
```
since you are only looking for `root_server_uri`, limit to just returning that field to keep performance optimal.

radius314 commented on endpoints/_includes/functions.php:
```diff
@@ -746,6 +748,30 @@ function helplineSearch($latitude, $longitude)
+function getBMLTserver($latitude, $longitude, $search_state)
+{
+	$helpline_search_radius=setting('helpline_search_radius');
+	if ($search_state=="1") {
```
Should use integers instead of strings.  More importantly, what does 1 and 2 mean?

spiritualtheory replied:
Switching to SearchType class

radius314 commented on endpoints/_includes/functions.php:
```diff
@@ -746,6 +748,30 @@ function helplineSearch($latitude, $longitude)
+				return $GLOBALS['bmlt_root_server'];
+			}
+		}
+	return getHelplineBMLTRootServer();
```
wrap in an `else` clause to make it more readable.

spiritualtheory replied:
Will move it up into the volunteer search section and add additional else. The reason for putting it outside of the if structure was because in the case of helpline volunteer, regardless of tomato on or off, the return was the same and the only difference was changing the tomato_helpline_enabled setting. Original thought was simpler code. Changed as requested.

radius314 commented on endpoints/_includes/functions.php:
```diff
@@ -759,7 +785,7 @@ function getFormatString($formats, $ignore = false, $helpline = false)
 function meetingSearch($meeting_results, $latitude, $longitude, $day)
 {
-    $bmlt_base_url = $GLOBALS['bmlt_root_server'] . "/client_interface/json/?switcher=GetSearchResults&data_field_key=meeting_name,weekday_tinyint,start_time,location_text,location_info,location_municipality,location_province,location_street,longitude,latitude,distance_in_miles,distance_in_km";
+    $bmlt_base_url = getBMLTserver($latitude, $longitude, 2) . "/client_interface/json/?switcher=GetSearchResults&data_field_key=meeting_name,weekday_tinyint,start_time,location_text,location_info,location_municipality,location_province,location_street,longitude,latitude,distance_in_miles,distance_in_km";
```
rename function to `getBMLTServer`

radius314 commented on endpoints/_includes/functions.php:
```diff
@@ -827,7 +853,8 @@ function getResultsString($filtered_list)
 function getServiceBodyCoverage($latitude, $longitude)
 {
-    $search_results = helplineSearch($latitude, $longitude);
+	getBMLTserver($latitude, $longitude, 1);
```
spacing issues need to fix to keep consistency with formatting

spiritualtheory replied:
Looking like a possible tab vs spacebar autoformat issue? They align evenly on the editor I used. Will run through another editor to see.

radius314 commented on endpoints/_includes/functions.php:
```diff
@@ -51,6 +51,8 @@
     'toll_province_bias' => [ 'description' => '' , 'default' => null, 'overridable' => true, 'hidden' => false],
     'toll_free_province_bias' => [ 'description' => '' , 'default' => '', 'overridable' => true, 'hidden' => false],
     'tomato_helpline_routing' => [ 'description' => '', 'default' => false, 'overridable' => true, 'hidden' => false],
+	'tomato_helpline_fallback' => [ 'description' => '', 'default' => false, 'overridable' => true, 'hidden' => false],
```
formatting issues with spaces needs to be fixed

spiritualtheory replied:
Possible tab vs space difference as noted above - they were otherwise cut/pasted from tomato_helpline_routing and just had the text changed for the setting.

spiritualtheory replied:
Was using notepad++ for quick editing - it was saving in a format it was not displaying. Everything looked well aligned on there until I looked at it with other editors. Used the basic notepad as a catch all since it has no formatting ability and everything looked aligned in there now.

radius314 commented on endpoints/_includes/functions.php:
```diff
@@ -746,6 +748,30 @@ function helplineSearch($latitude, $longitude)
     return json_decode(get($search_url));
 }
 
+function getBMLTserver($latitude, $longitude, $search_state)
```
There are a lot of formatting issues going on here.  There should be spaces between equal signs, periods for string concatenation (`=`, `.`).  curly brackets are not lining up like the rest of the code base.

spiritualtheory commented:
Think I cleaned up what was commented/requested. Initial change to using the SearchType class I was hitting 3 : 's instead of 2 and pushed before catching it. Fixed. Tested on NNY's server as working.

radius314 commented:
You should either install the notepad++ editor config plugin, or use something like atom or sublime text which automatically support editorconfig conventions.

spiritualtheory commented:
Actually I did the work on notepad++

However, you pointed out formatting issues and when I opened it in regular notepad, I could see the issues. Will have to see what the glitch was - why it formatted as it did.

radius314 commented:
Regular notepad probably doesn’t support editorconfig.

On Fri, Nov 15, 2019 at 11:40 AM spiritualtheory <notifications@github.com>
wrote:

> Actually I did the work on notepad++
>
> However, you pointed out formatting issues and when I opened it in regular
> notepad, I could see the issues. Will have to see what the glitch was - why
> it formatted as it did.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/bmlt-enabled/yap/pull/326?email_source=notifications&email_token=AAOD3O3NZ5C2VORECIRNTFLQT3GHRA5CNFSM4JNMGQ2KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEGADZI#issuecomment-554435045>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOD3OY5TTP23VRMYTD422TQT3GHRANCNFSM4JNMGQ2A>
> .
>

radius314 commented:
Notepad++ doesn’t support editorconfig by default.  My suggestion is to use a proper ide.  Visual studio code or IntelliJ.

spiritualtheory commented:
Yes, I know - however you pointed out the spacing being bad - when I looked at it notepad++ it looked fine - everything was aligned as it should except for some misplaced } 's - I saw no issue with the spacing. When I opened in notepad, the function I created was skewed out of proportion. So for the sake of just making sure to get it done quickly and surely, I just fixed it in notepad.

spiritualtheory commented:
I've been looking at a few IDE's - not impressed with the few I've checked out - will check out and see where Visual Studio is now as well as the other you suggested. Nostalgia for me, last time I worked with Visual Studio was early 2000's with C++ with CPU/hardware emulation.

radius314 commented:
Most engineers on the BMLT project use IntelliJ.  Visual Studio Code is
nothing like the old Visual Studio.  It’s very lightweight.

On Fri, Nov 15, 2019 at 11:45 AM spiritualtheory <notifications@github.com>
wrote:

> I've been looking at a few IDE's - not impressed with the few I've checked
> out - will check out and see where Visual Studio is now as well as the
> other you suggested. Nostalgia for me, last time I worked with Visual
> Studio was early 2000's with C++ with CPU/hardware emulation.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/bmlt-enabled/yap/pull/326?email_source=notifications&email_token=AAOD3OYUPPQYEVWTNMU5SI3QT3G4FA5CNFSM4JNMGQ2KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEGATKI#issuecomment-554437033>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOD3O6ZAVB6AOCMQDTJQGTQT3G4FANCNFSM4JNMGQ2A>
> .
>

spiritualtheory commented:
Nice - I will definitely check out IntelliJ - I'd like to help with bits and pieces here and there where I can.

pjaudiomv commented:
So yap has an `.editorconfig` https://github.com/bmlt-enabled/yap/blob/master/.editorconfig  file this defines and helps maintain consistent coding styles between different editors and IDEs. Whatever editor you use is purely preference however I an would highly suggest using one that supports [editorconfig](https://editorconfig.org/). this makes contributing to projects much easier.

pjaudiomv commented:
I use intellij and use BBEdit for my lightweight editor

spiritualtheory commented:
Sounds good - like I said, I'd like to contribute as I can. 

This piece really was just thinking about the use case of a NEZF Phone Line with 3 root servers, how to manage and not break YAP admin or volunteer settigns - and not have to convince 3 root server admin regions to allow the zone to take those over. Also thought about San Francisco area currently sitting on KY's server with no visibility of anything else in Cali (until WSZF BMLT goes and stays live - lol) - and can't break their yap admin panel to create visibility. 

So I took a few hours one night this week to figure out the flow and workings of YAP (broad view) to see how I could fix that issue without breaking anything else.  I looked at a few IDE's, didn't like them, settled on notepad++ and went from there. Outside of commenting out the JFT bug I pointed out last year (since fixed), this was my first foray into PHP at all and my first actual coding (beyond VBA/Excel) in about 16 years. Felt good.

I look forward to seeing where else I can possibly contribute as well as learning along the way from those far far more experienced than I.

radius314 commented on endpoints/_includes/functions.php:
```diff
@@ -827,6 +855,8 @@ function getResultsString($filtered_list)
 
 function getServiceBodyCoverage($latitude, $longitude)
 {
+    getBMLTServer($latitude, $longitude, SearchType::VOLUNTEERS);
```
Should we be doing something with the output from this function?

spiritualtheory replied:
At that particular location no. That is just to run a check on conditions to see if tomato_helpline_routing is needed to be flipped on or not. So the return output from the function is not needed there.

radius314 replied:
A function should do one thing, and do it well.  The way I'm reading this code, the function does two things, returns a value and also changes a state from inside the function.  My suggestion is to decouple these behaviors.

spiritualtheory replied:
I can break out the rootserver verification into a separate function easily enough.

radius314 commented on endpoints/_includes/functions.php:
```diff
@@ -746,6 +748,32 @@ function helplineSearch($latitude, $longitude)
+            }
+            return getHelplineBMLTRootServer();
+        } else {
+            return getHelplineBMLTRootServer();
```
seems like both these conditions return the same thing, what's the difference?

spiritualtheory replied:
There isn’t - I originally had the return at the end outside of the if statement because of that, but you had said wrap it in and else statement for readability so I did.

The return doesn’t need to be inside the if because the only thing that portion of the if does is enable the setting tomato_helpline_routing to true if it is a volunteer search and tomato fallback is requested and the rootserver handling the routing is not the one the yap server is configured to.

I can move that back outside of the if statement where it originally was.

radius314 replied:
I guess I didn't mean for it to wind up here.  But I also see that there is not always an exiting return happening, so it makes a little more sense now.

radius314 commented on endpoints/_includes/functions.php:
```diff
@@ -746,6 +748,32 @@ function helplineSearch($latitude, $longitude)
+            $pinged_BMLT_root = $search_results[0] -> root_server_uri;
+            $bmlt_root = $GLOBALS['bmlt_root_server']."/";
+            if ($pinged_BMLT_root != $bmlt_root) { // If the rootserver handling the phonecall is not the local rootserver, enable tomato routing - otherwise handle call routing locally.
+                $GLOBALS['settings_whitelist']['tomato_helpline_routing']['default'] = true;
```
Btw, this isn't going to change anything, I do not recommend modifying the behavior of settings in this manner.

spiritualtheory replied:
Changing the tomato_helpline_routing from within the instance allows it to simply flip a switch that the rest or the already built code can work with. Keeps it unobtrusive to the existing code base. Would require more extensive rewriting to otherwise accomplish the same thing.

So would this mean that even once the above requests are satisfied this will be rejected for the menthod? Or are you saying this won’t work? It has tested out working for all possible logic conditions.

radius314 replied:
The problem with this is that it's not embedded in the session, so the behavior will be lost on subsequent calls if any.  This is not a code pattern that exists anywhere else and may cause unwanted side effects.

spiritualtheory replied:
It is intentionally embedded in the session - it is a temporary flipping of the switch for the purpose of the session it is inside and all references to tomato_helpline_routing - so that no other code needed to be modified. It only flips that switch under the following conditions : volunteer search + tomato_helpline_fallback true + compare target rootserver vs config rootserver is different. For all 3 of those to be true is a singular set of conditions that prevents interference with other functions.

radius314 commented:
I wouldn't spin your wheels on this too much, I can take over the work on this, as use this as inspiration.

spiritualtheory commented:
Rewrote the solution to accommodate not working with flipping global switches, even in session.

Moved the function out for comparing what rootserver is local to the requested location against the rootserver in the global config

Moved everything else into the original getHelplineBMLTRootServer. In the case of null arguments, it passes through and behaves identical to original function. In the case of being called for meeting search or looking for helpline, it checks conditions for tomato settings and usefulness.

I could see putting the null check ahead of the search type check since most calls to the function will be null and that would save a couple instructions for performance sake. Put it at the end as a catch all for undefined or null arguments to keep behavior simple and readable.

As a side note - love IntelliJ - still getting used to it, but enjoying it.

radius314 commented:
There is actually another way to potentially handle this.  Yap 3.5.0 allows you to define users outside of BMLT users.  What this means is that you can tie users to service body ideas in Tomato and set up all kind of interesting routing rules.  

Also I just pushed a change that removes the semanticAdmin check which was no longer in use (phased out in Yap 3.x).

spiritualtheory commented:
That is an interesting additional feature that I like too

But a separate one from simply setting up tomato fallback.

Ran into an admin bug on the last code - working on that at the moment

radius314 commented:
That bug has fixed, it was related to checking for semanticAdmin flag.

radius314 commented on endpoints/_includes/functions.php:
```diff
@@ -746,6 +748,42 @@ function helplineSearch($latitude, $longitude)
+        if (setting('tomato_helpline_fallback') & !BMLTServerIsLocalConfig($latitude, $longitude)) {
+            return setting('tomato_url');
+        }
+    } else if (($search_type == SearchType::MEETINGS) && setting('tomato_meeting_search')) {
```
This pull request is almost good to go.  Why are we coupling meeting searches with a function called `getHelplineBMLTRootServer`? 

Secondly, Isn't using the tomato url deliberate enough to say use tomato rather than add a setting?

spiritualtheory replied:
Just to unify the getting of the BMLT root server

And finding that some piece in the most recent commit broke the admin panel and local yap volunteer routing - all the settings on/off work and tested to know what rootserver to hit - but the getservicebodies is not functioning correctly. Some of my assumptions on null arguments seem to be incorrect.

spiritualtheory replied:
As to the tomato setting - just separated out whether someone wants tomato fallback on their helpline versus meeting searches - which could be separated.

radius314 replied:
If that's the case it should be called `getBMLTRootServer()`

There should be no change that broke the admin panel.

radius314 replied:
The setting of `$helpline_bmlt_root_server` and `$bmlt_root_server` is the distinction between meeting searches and helpline lookups.  I"m not understanding why that is not sufficient.

radius314 replied:
I do not see `tomato_meeting_search` as a change that is necessary.  You can choose to use whatever root server you want by setting it.  If you set `helpline_bmlt_root_server` it becomes mutually exclusive.

radius314 commented:
There is some breaking code here.  I am going to be making some changes.

spiritualtheory commented:
Yeah, I know

The version at the beginning of the day worked without breaking admin

I didn't test admin again until the last commit - just the conditions of yap's functionality

Also working on sorting out where the break is.
